### ISSUE TITLE
DiscoveryWithServiceDisruptionsIT#testAckedIndexing fails with NoShardAvailableActionException

### ISSUE THREAD

**Issue by danielmitterdorfer** [elastic/elasticsearch#24457]
* Failure link: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.x+multijob-unix-compatibility/os=ubuntu/755/consoleFull
* Reproduction line: `gradle :core:integTest -Dtests.seed=C250F03598055A69 -Dtests.class=org.elasticsearch.discovery.DiscoveryWithServiceDisruptionsIT -Dtests.method="testAckedIndexing" -Dtests.security.manager=true -Dtests.locale=ar-AE -Dtests.timezone=Pacific/Fakaofo`
* Failure rate: failed 12 times in CI

<details>
    <summary>Error trace</summary>
    <p><pre>NoShardAvailableActionException[No shard available for [get [test][type][11]: routing [null]]]
   > 	at __randomizedtesting.SeedInfo.seed([C250F03598055A69:489144C6C478BC22]:0)
   > 	at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction.perform(TransportSingleShardAction.java:209)
   > 	at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction.start(TransportSingleShardAction.java:186)
   > 	at org.elasticsearch.action.support.single.shard.TransportSingleShardAction.doExecute(TransportSingleShardAction.java:95)
   > 	at org.elasticsearch.action.support.single.shard.TransportSingleShardAction.doExecute(TransportSingleShardAction.java:59)
   > 	at org.elasticsearch.action.support.TransportAction.doExecute(TransportAction.java:146)
   > 	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:170)
   > 	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:142)
   > 	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:84)
   > 	at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:83)
   > 	at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:72)
   > 	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:408)
   > 	at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:80)
   > 	at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:54)
   > 	at org.elasticsearch.action.ActionRequestBuilder.get(ActionRequestBuilder.java:62)
   > 	at org.elasticsearch.discovery.DiscoveryWithServiceDisruptionsIT.testAckedIndexing(DiscoveryWithServiceDisruptionsIT.java:609)
   > 	at java.lang.Thread.run(Thread.java:745)

</pre></p>
</details>

[relevant part from the build log](https://github.com/elastic/elasticsearch/files/972958/build_log.txt)

**Comment by ywelsch** [comment]
TL;DR: The cluster health request with local flag set to true is super trappy

What happens here is the following:
1) at the end of the test, the disruption is stopped
```
  1> [2017-05-03T22:08:49,567][INFO ][o.e.d.DiscoveryWithServiceDisruptionsIT] stopping disruption
  1> [2017-05-03T22:08:49,567][INFO ][o.e.t.d.NetworkDisruption] stop disrupting (disruption scheme: network unresponsive, disrupted links: two partitions (partition 1: [node_t2] and partition 2: [node_t1, node_t0]))
```
2) the test then calls `ensureStableCluster` for each node and also `ensureGreen`
```
  1> [2017-05-03T22:08:49,568][DEBUG][o.e.d.DiscoveryWithServiceDisruptionsIT] ensuring cluster is stable with [3] nodes. access node: [node_t0]. timeout: [40s]
  1> [2017-05-03T22:08:49,568][TRACE][o.e.c.s.ClusterService   ] [node_t0] will process [cluster_health (wait_for_events [LANGUID])]
  1> [2017-05-03T22:08:49,568][DEBUG][o.e.c.s.ClusterService   ] [node_t0] processing [cluster_health (wait_for_events [LANGUID])]: execute
  1> [2017-05-03T22:08:49,568][DEBUG][o.e.c.s.ClusterService   ] [node_t0] processing [cluster_health (wait_for_events [LANGUID])]: took [0s] no change in cluster_state
  1> [2017-05-03T22:08:49,568][DEBUG][o.e.d.DiscoveryWithServiceDisruptionsIT] ensuring cluster is stable with [3] nodes. access node: [node_t1]. timeout: [40s]
  1> [2017-05-03T22:08:49,569][TRACE][o.e.c.s.ClusterService   ] [node_t1] will process [cluster_health (wait_for_events [LANGUID])]
  1> [2017-05-03T22:08:49,569][DEBUG][o.e.c.s.ClusterService   ] [node_t1] processing [cluster_health (wait_for_events [LANGUID])]: execute
  1> [2017-05-03T22:08:49,569][DEBUG][o.e.c.s.ClusterService   ] [node_t1] processing [cluster_health (wait_for_events [LANGUID])]: took [0s] no change in cluster_state
  1> [2017-05-03T22:08:49,569][DEBUG][o.e.d.DiscoveryWithServiceDisruptionsIT] ensuring cluster is stable with [3] nodes. access node: [node_t2]. timeout: [40s]
  1> [2017-05-03T22:08:49,569][TRACE][o.e.c.s.ClusterService   ] [node_t2] will process [cluster_health (wait_for_events [LANGUID])]
  1> [2017-05-03T22:08:49,569][DEBUG][o.e.c.s.ClusterService   ] [node_t2] processing [cluster_health (wait_for_events [LANGUID])]: execute
  1> [2017-05-03T22:08:49,569][DEBUG][o.e.c.s.ClusterService   ] [node_t2] processing [cluster_health (wait_for_events [LANGUID])]: took [0s] no change in cluster_state
  1> [2017-05-03T22:08:49,570][TRACE][o.e.c.s.ClusterService   ] [node_t0] will process [cluster_health (wait_for_events [LANGUID])]
  1> [2017-05-03T22:08:49,570][DEBUG][o.e.c.s.ClusterService   ] [node_t0] processing [cluster_health (wait_for_events [LANGUID])]: execute
  1> [2017-05-03T22:08:49,571][DEBUG][o.e.c.s.ClusterService   ] [node_t0] processing [cluster_health (wait_for_events [LANGUID])]: took [0s] no change in cluster_state
  1> [2017-05-03T22:08:49,571][DEBUG][o.e.d.DiscoveryWithServiceDisruptionsIT] indices [test] are green
```
3) finally the test does a number of `GET` operations to validate that all acked operations have been successfully indexed.

The issue is that `ensureStableCluster` and `ensureGreen` get to execute and properly return while the cluster is still in a transition state. `node_t2` in particular is still isolated from the rest of the cluster and has a state with a `NO_MASTER_BLOCK` but the cluster health action with local flag that is done as part of `ensureStableCluster` does not care about that. It happily reports health as green 👎 
Nothing in the health response indicates that the node that was queried is not part of a cluster.

/cc: @bleskes @jasontedor

**Comment by bleskes** [comment]
@ywelsch I think the idea of the local flag is to make sure that all nodes have processed the fact that the cluster is stable. I think a better way is to first reach out to the master and then validate all nodes have applied at least as high a cluster state version. Does that make sense?

**Comment by bleskes** [comment]
@ywelsch do you still feel that we should pick this up? if so, what do you think of my suggestion?

**Comment by mayya-sharipova** [comment]
failures still continue to occasionally happen on 5.6 with a rate 1-2 times a week:

Recent failure

Log: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.6+multijob-unix-compatibility/os=sles/33/console

**Comment by henningandersen** [comment]
We still have failures from this test though with different errors, for instance this one:

https://scans.gradle.com/s/npcyedhjecs2m/tests/kyv2y2z3r4v7m-f6hbue3yhtsqi?openStackTraces=WzBd

A notable warning here is:

```
[2019-06-15T03:48:31,786][WARN ][o.e.t.n.MockNioTransport ] [node_t2] Slow execution on network thread [49634 milliseconds]
java.lang.RuntimeException: Slow exception on network thread
	at org.elasticsearch.transport.nio.MockNioTransport$TransportThreadWatchdog.maybeLogElapsedTime(MockNioTransport.java:367) ~[framework-7.2.0-SNAPSHOT.jar:7.2.0-SNAPSHOT]
	at org.elasticsearch.transport.nio.MockNioTransport$TransportThreadWatchdog.unregister(MockNioTransport.java:359) ~[framework-7.2.0-SNAPSHOT.jar:7.2.0-SNAPSHOT]
	at org.elasticsearch.transport.nio.TestEventHandler.acceptChannel(TestEventHandler.java:54) ~[framework-7.2.0-SNAPSHOT.jar:7.2.0-SNAPSHOT]
	at org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:227) ~[elasticsearch-nio-7.2.0-SNAPSHOT.jar:7.2.0-SNAPSHOT]
	at org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:172) ~[elasticsearch-nio-7.2.0-SNAPSHOT.jar:7.2.0-SNAPSHOT]
	at org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:129) ~[elasticsearch-nio-7.2.0-SNAPSHOT.jar:7.2.0-SNAPSHOT]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
```

which does look similar to #43387. Backported #42617 to 7.2 to maybe get a bit more info out if this happens again.

**Comment by henningandersen** [comment]
Checked all the recent failures of this test and they fall into following categories:

1. Network unresponsive disruption, none of those since #42579 was merge, though this is only 9 days ago (so need more time to be conclusive on that part).
2. Global checkpoint not in sync, should be fixed by #41068, now in 7.2 too.
3. The above symptom (twice in the last week, both on 7.2, once on debian 8 and once on zulu11/docker build).

**Comment by henningandersen** [comment]
New incident from 7.x, same symptom, but with the additional logging that @original-brownbear added:

```
[2019-06-20T10:24:56,622][WARN ][o.e.t.n.MockNioTransport ] [node_t0] Potentially blocked execution on network thread [elasticsearch[node_t0][transport_worker][T#2]] [55630 milliseconds]: 
sun.nio.ch.ServerSocketChannelImpl.accept0(Native Method)
sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:422)
sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:250)
org.elasticsearch.nio.ChannelFactory$RawChannelFactory$$Lambda$1613/1836336234.run(Unknown Source)
java.security.AccessController.doPrivileged(Native Method)
org.elasticsearch.nio.ChannelFactory$RawChannelFactory.accept(ChannelFactory.java:223)
org.elasticsearch.nio.ChannelFactory$RawChannelFactory.acceptNioChannel(ChannelFactory.java:180)
org.elasticsearch.nio.ChannelFactory.acceptNioChannel(ChannelFactory.java:55)
org.elasticsearch.nio.ServerChannelContext.acceptChannels(ServerChannelContext.java:47)
org.elasticsearch.nio.EventHandler.acceptChannel(EventHandler.java:45)
org.elasticsearch.transport.nio.TestEventHandler.acceptChannel(TestEventHandler.java:51)
org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:227)
org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:172)
org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:129)
org.elasticsearch.nio.NioSelectorGroup$$Lambda$1556/832165726.run(Unknown Source)
java.lang.Thread.run(Thread.java:748)
[2019-06-20T10:24:56,699][WARN ][o.e.t.n.MockNioTransport ] [node_t1] Potentially blocked execution on network thread [elasticsearch[node_t1][transport_worker][T#2]] [56029 milliseconds]: 
sun.nio.ch.ServerSocketChannelImpl.accept0(Native Method)
sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:422)
sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:250)
org.elasticsearch.nio.ChannelFactory$RawChannelFactory$$Lambda$1613/1836336234.run(Unknown Source)
java.security.AccessController.doPrivileged(Native Method)
org.elasticsearch.nio.ChannelFactory$RawChannelFactory.accept(ChannelFactory.java:223)
org.elasticsearch.nio.ChannelFactory$RawChannelFactory.acceptNioChannel(ChannelFactory.java:180)
org.elasticsearch.nio.ChannelFactory.acceptNioChannel(ChannelFactory.java:55)
org.elasticsearch.nio.ServerChannelContext.acceptChannels(ServerChannelContext.java:47)
org.elasticsearch.nio.EventHandler.acceptChannel(EventHandler.java:45)
org.elasticsearch.transport.nio.TestEventHandler.acceptChannel(TestEventHandler.java:51)
org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:227)
org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:172)
org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:129)
org.elasticsearch.nio.NioSelectorGroup$$Lambda$1556/832165726.run(Unknown Source)
java.lang.Thread.run(Thread.java:748)
[2019-06-20T10:24:56,761][WARN ][o.e.t.n.MockNioTransport ] [node_t2] Potentially blocked execution on network thread [elasticsearch[node_t2][transport_worker][T#2]] [54829 milliseconds]: 
sun.nio.ch.ServerSocketChannelImpl.accept0(Native Method)
sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:422)
sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:250)
org.elasticsearch.nio.ChannelFactory$RawChannelFactory$$Lambda$1613/1836336234.run(Unknown Source)
java.security.AccessController.doPrivileged(Native Method)
org.elasticsearch.nio.ChannelFactory$RawChannelFactory.accept(ChannelFactory.java:223)
org.elasticsearch.nio.ChannelFactory$RawChannelFactory.acceptNioChannel(ChannelFactory.java:180)
org.elasticsearch.nio.ChannelFactory.acceptNioChannel(ChannelFactory.java:55)
org.elasticsearch.nio.ServerChannelContext.acceptChannels(ServerChannelContext.java:47)
org.elasticsearch.nio.EventHandler.acceptChannel(EventHandler.java:45)
org.elasticsearch.transport.nio.TestEventHandler.acceptChannel(TestEventHandler.java:51)
org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:227)
org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:172)
org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:129)
org.elasticsearch.nio.NioSelectorGroup$$Lambda$1556/832165726.run(Unknown Source)
java.lang.Thread.run(Thread.java:748)
```

https://scans.gradle.com/s/62jsofae5l5n4/tests/kyv2y2z3r4v7m-f6hbue3yhtsqi

**Comment by henningandersen** [comment]
Just realized that all 3 above build failures and the one in #43387 have run on debian 8. That runs kernel 3.16, which is subject to:

https://groups.google.com/forum/m/#!msg/mechanical-sympathy/QbmpZxp6C64/HRwMTvPXl1UJ

fixed in 3.18 but also in a number of point fixes. I am going to chase that direction a bit more.

**Comment by henningandersen** [comment]
The futex wait should be fixed long ago in that kernel:

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-3.16.y&id=b89814999ab2a2120be803c1468fc54ad965df1f

Tried to reproduce the accept hang issue in a minimalistic program doing connects and disconnects, but were not able to provoke it.

**Comment by original-brownbear** [comment]
@henningandersen same here, I ran a bunch of rounds of the failing test task over night (whole Gradle run) on Debian 8 without any hanging. Might not be the futex issue, but certainly is something else system specific.

**Comment by ywelsch** [comment]
The remaining problems are tracked in https://github.com/elastic/elasticsearch/issues/41068 and https://github.com/elastic/elasticsearch/issues/43387
